baseband_tasks/dispersion.py

```diff
@@ -182,3 +182,44 @@ def __init__(self, ih, dm, reference_frequency=None,
     @property
     def dm(self):
         return -self._dm
+
+
+class DisperseSamples(SampleShift):
+    """Incoherently shift a time stream based on the disperse time delay.
+
+    This task does not handle the in channel smearing, only shift the samples.
+
+    Parameters
+    ----------
+    ih : task or `baseband` stream reader
+        Input data stream, with time as the first axis.
+    dm : float or `~baseband_tasks.dm.DispersionMeasure` quantity
+        Dispersion measure.
+    reference_frequency : `~astropy.units.Quantity`
+        Frequency to which the data should be dedispersed.  Can be an array.
+        By default, the mean frequency.  If one doesn't want to change the
+        start time, choose the maximum frequency.
+    samples_per_frame : int, optional
+        Number of dedispersed samples which should be produced in one go.
+        The number of input samples used will be larger to avoid wrapping.
+        If not given, as produced by the minimum power of 2 of input
+        samples that yields at least 75% efficiency.
+    frequency : `~astropy.units.Quantity`, optional
+        Frequencies for each channel in ``ih`` (channelized frequencies will
+        be calculated).  Default: taken from ``ih`` (if available).
+    """
+    def __init__(self, ih, dm, reference_frequency=None,
+                 samples_per_frame=None, frequency=None):
+        # Compute the time shift
+        # TODO: 1. How to remove the duplicated code
+        # 2. Is calculating delay based on the input frequency the right thing
+        # to do?
+        dm = DispersionMeasure(dm)
+        # The sample disperse does not handle the sidebands, since it does not
+        # handle the in channel smearing.
+        frequency = getattr_if_none(ih, 'frequency', frequency)
+        reference_frequency = frequency.mean()
+
+        # Treat the input frequency as the time delay frequency.
+        time_delay = dm.time_delay(freqency, reference_frequency)
+        super().__init__(ih, time_delay, samples_per_frame=samples_per_frame)
```

Comment:
Should I pass the "frequency" to the SampleShift class? I am not sure how does the metadata flow.

Reply:
Hmm, tricky one. Really here the frequencies are only used to calculate the offsets. But generally, the idea is that once you pass in frequencies, then the task itself will store that in `.frequency` so that you do not have to pass it again for another task. But that also means one has to pass in `sideband` as well, since the code will not allow one without the other. And `ShiftSamples` has to be adjusted to take a `**kwargs` to pass on to `PaddedTaskBase`...

I guess the alternative is not to let the user pass in frequency at all, but that seems inconsistent with the other dispersion tasks (though perhaps is worth doing as a larger refactoring).

Yet another alternative is
```
if frequency is not None or sideband is not None:
    ih = SetAttribute(ih, frequency=frequency, sideband=sideband)
```
and then continue as you have it (now knowing `ih.frequency` will exist).

That leaves `ShiftSamples` untouched so is perhaps the best solution.

Reply:
Saw your change.  I meant further up but will just do a quick commit myself, also cleaning up some other stuff.